qutebrowser/components/braveadblock.py

```diff
@@ -0,0 +1,313 @@
+        self._wip_filter_set = adblock.FilterSet()
+        logger.info("Downloading adblock filter lists...")
+
+        blocklists = config.val.content.blocking.adblock.lists
+        if not blocklists:
+            # Blocklists are None or length zero
+            self._on_lists_downloaded()
+        else:
+            self._finished_registering_downloads = False
+            for i, url in enumerate(blocklists):
+                if i == len(blocklists) - 1:
+                    self._finished_registering_downloads = True
+                blockutils.download_blocklist_url(
+                    url, self._on_download_finished, self._in_progress
+                )
+
+    def _on_lists_downloaded(self) -> None:
+        """Install block lists after files have been downloaded."""
+        assert self._wip_filter_set is not None
+        self._engine = adblock.Engine(self._wip_filter_set)
+        self._wip_filter_set = None
+        self._engine.serialize_to_file(str(self._cache_path))
+        logger.info(
+            "adblock: Filters successfully read from {} sources".format(
+                self._done_count
+            )
+        )
+
+    def update_files(self) -> None:
+        """Update files when the config changed."""
+        if not config.val.content.blocking.adblock.lists:
+            try:
+                os.remove(self._cache_path)
+            except FileNotFoundError:
+                pass
+            except OSError as e:
+                logger.exception("Failed to remove adblock cache file: {}".format(e))
+
+    def _on_download_finished(self, download: downloads.TempDownload) -> None:
+        """Check if all downloads are finished and if so, trigger reading.
+
+        Arguments:
+            download: The finished download.
+        """
+        self._in_progress.remove(download)
+        if download.successful:
+            self._done_count += 1
+            assert not isinstance(download.fileobj, downloads.UnsupportedAttribute)
+            assert download.fileobj is not None
+            assert self._wip_filter_set is not None
+            try:
+                download.fileobj.seek(0)
+                text = io.TextIOWrapper(download.fileobj, encoding="utf-8")
+                self._wip_filter_set.add_filter_list(text.read())
+                text.close()
+            except UnicodeDecodeError:
+                message.info("Block list is not valid utf-8")
+            finally:
+                download.fileobj.close()
+        if len(self._in_progress) == 0 and self._finished_registering_downloads:
+            try:
+                self._on_lists_downloaded()
+            except OSError:
+                logger.exception("Failed to write host block list!")
+
+
+@hook.config_changed("content.blocking.adblock.lists")
+def on_lists_changed() -> None:
+    if ad_blocker is not None:
+        ad_blocker.update_files()
+
+
+@hook.config_changed("content.blocking.method")
+def on_method_changed():
```

Comment:
Could the reason for the failing test `test_blocking_enabled[True, "hosts", False]` be that this hook doesn't trigger for some reason?

Reply:
Yep! The `@hook.…` decorators only [register those hooks](https://github.com/qutebrowser/qutebrowser/blob/v1.13.1/qutebrowser/api/hook.py#L89-L92) - they then get [wired up](https://github.com/qutebrowser/qutebrowser/blob/v1.13.1/qutebrowser/extensions/loader.py#L169-L182) by the extension loader. If you import the module inside pytest without the loader involved, that won't happen.

In the long run, there should probably some easy way to load an extension module from tests - but for now, it's probably easiest if you just call the hook manually from the test.